packages/datetime/src/dateInput.tsx

```diff
@@ -298,7 +298,9 @@ export class DateInput extends AbstractComponent<IDateInputProps, IDateInputStat
     private handleInputBlur = () => {
         const valueString = this.state.valueString;
         let value = moment(valueString, this.props.format);
-        if (valueString !== this.getDateString(this.state.value) && (!value.isValid() || !this.dateIsInRange(value))) {
+        if (valueString.length &&
```

Comment:
nit: we prefer putting the binary operator on the next line for multiline expressions, like this:

```ts
if (valueString.length
    && valueString ...
```

(yes, this really should be enforced by our linter)